✨Windows Support

cowboyd opened the pull request:
> resolves #432

Motivation
-----------

Bigtest development workflows should work on windows as well. This will not only guarantee that bigtest can run on windows, but that BigTest itself can be developed on Windows whether a developer is using cmd.exe, powershell.exe, or the linux subsystem.

Approach
----------

Most of the problems with running the tests on windows have to do with the build scripts not being able to run. For example, `chmod`, `rm -rf`, etc are unix-only commands. Also, the trick of setting a one-off environment variable does not work for running a command (FOO=bar my-command).

In order to make these robustly cross-platform, build scripts have been moved, where necessary into javascript.

cowboyd committed:
Convert todomvc to nodejs script in order to make it cross platform

The todomvc build fails on windows because it uses a string of
unix-only commands. This causes the entire build to fail as a result
on Windows.

This translates the command string into a build.js script that just
does the same sequence of commands except in node.

cowboyd committed:
Convert server generate schema script to JS

It's hard to set environment variables on a per-command basis from the
command prompt in a cross-platform way, so this converts it to a JS script

cowboyd committed:
unify @effection/node version at 0.9.1

v0.9.1 was the version that introduced windows support.

cowboyd committed:
pre-munge ts-node path so that it works with exec()

Our exec() command uses shellwords to split command and
arguments. However, it has a path sepearator bug that fails to parse
absolute paths. This "pre-munges" the `ts-node` command into unix
format so that it can be un-munged by cross-spawn. A hack for sure,
but without an upstream fix on the `shellwords` package. That's where
we are.

cowboyd committed:
Build CLI with cross platform scripts

cowboyd committed:
win32 takes longer, *shrug*

cowboyd committed:
only assert on a portion of the source line

In windows, the source path has `\` characters, and only the end can
be asserted for some reason.

cowboyd committed:
Enable Tests on Windows

cowboyd committed:
effection node deps are actually a runtime dep

cowboyd committed:
Add windows to the test matrix

cowboyd committed:
Add platform to test name

jbolda commented on packages/agent/test/agent.test.ts:
```diff
@@ -32,7 +32,12 @@ describe("@bigtest/agent", function() {
 
-  this.timeout(process.env.CI ? 60000 : 10000);
+  if (process.platform === 'win32') {
+    this.timeout(process.env.CI ? 120000 : 30000);
```
😂

cowboyd committed:
upgrade chromedriver and use exe files directly

jnicklas commented on packages/cli/package.json:
```diff
@@ -16,7 +16,7 @@
     "mocha": "mocha -r ts-node/register",
     "bigtest": "ts-node src/index.ts",
-    "prepack": "tsc --outdir dist --module commonjs && chmod +x dist/bin/bigtest.js"
+    "prepack": "tsc --outdir dist --module commonjs && node chmod.js 755 dist/bin/bigtest.js"
```
Why do we even need to `chmod` here? Shouldn't yarn/npm take care of this for us?

dagda1 replied:
You will need `chmod` for shebangs etc.

you can  use the `--chmod` switch of `git update-index` though to commit

e.g.

```bash
git update-index --chmod=755 --add dist/bin/bigtest.js
```

cowboyd replied:
Never new that was a even a thing!

jnicklas commented on packages/server/generate-schema.js:
```diff
@@ -0,0 +1,52 @@
+  console.log(`$ ${command}`);
+  let child = yield exec(command, options);
+
+  yield spawn(child.stdout.subscribe().forEach(function*(datum) {
+    process.stdout.write(datum);
+  }));
+
+  yield spawn(child.stderr.subscribe().forEach(function*(datum) {
+    process.stderr.write(datum);
+  }));
```
Can't we just use `stdio: inherit`?

cowboyd replied:
We actually don't export the `stdio` option in your `exec()` api https://github.com/thefrontside/effection/blob/v0/packages/node/src/exec/api.ts#L25-L47

We could add that to `exec()`, but now that I think of it, that was always a pretty confusing API. Maybe we should add `sh()` to `@effection/node`?

jbolda replied:
If we "inherit", then we can't act on any of the output, essentially a pure side effect. Using "inherit" is what nearly every use of child_process does which is partly why we are likely one of the few that has run into the deep issues on windows that we have had. Maybe another function makes sense, but I wonder if the use case might be outside the scope of this package?

cowboyd replied:
Yeah, it's a dang useful function, but forcing the the user to decide where the bytes go feels like the right primitive.

cowboyd committed:
Get rid of out-dated build artifacts

cowboyd committed:
Skip manifest builder tests

It appears that the file system notifications are a bit wonky and when
using the server triggering rebuilds more often than is
necessary. This appears to result in flaky tests. Rather than hold up
this PR, we skip these tests on windows only.

jbolda reviewed (approved):
Looking forward to make use of this!

dagda1 reviewed (approved):
we can come back to this at any time.
